---
title: "Class 9: Structural Bioinformatics 1"
author: "Ian Gurholt (PID: A16767484)"
format: gfm
---

The main database for structural data is called PDB (Protein Databank), let's
see what it contains

Data from: https://www.rcsb.org/stats

```{r}
pdbdb<- read.csv("PDB_stats.csv",row.names=1)
pdbdb
```

```{r}
pdbdb$Total
```

I need to remove the commas and convert to numberic in order to do math:
```{r}
as.numeric(sub(",","", pdbdb$Total))
```
I could turn this into a function to fix the whole table or any future table I read like this.

```{r}
x<- pdbdb$Total
as.numeric(sub(",","", pdbdb$Total))
```

```{r}
comma2numeric<- function(x){
  as.numeric(sub(",","", x))
}
```

Test it

```{r}
comma2numeric(pdbdb$X.ray)
```
```{r}
apply(pdbdb, 2, comma2numeric)
```
## Or try a different read/import function:

```{r}
library(readr)
pdbdb<-read_csv("PDB_stats.csv")
pdbdb
```
> Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
sum(pdbdb$`X-ray`)/(sum(pdbdb$Total)) * 100
```


> Q2: What proportion of structures in the PDB are protein?

```{r}
percentage <- (pdbdb[1, 8] / sum(pdbdb$Total)) * 100
percentage
```
> Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

There are currently 4,563 structures of HIV-1 in PDB database. 


Mol* (pronounced “molstar”) is a new web-based molecular viewer that we will need to learn the basics of here

Accessed via: https://molstar.org/viewer/.

We will use PDB code: 1HSG

![A first image from molestar of HIV-1 protease](1HSG.png)

Some more images from molestar: 

![A second image from molestar with D25 amino acid displayed](1HSG (2).png)

![A third image from molestar with spacefill added to show tight binding sight with drug](1HSG (3).png)

> Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

We do not see the hydrogen atoms of the water molecules and only see the oxygen atoms because if we were to see all the atoms, including the hydrogen atoms on all the other molecules of the protein structure, it would make the visual way to crowded and complicated, obscuring the view of the important structures such as side chains and binding pockets. Furthermore, we can still identify bonds as the hydrogens contribute very little to the specific interactions formed within the protein structure. 

> Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

This is water molecule 308 and is critical within the binding site as it bind/stabilizes the ligand within the protein.

![An image identifying the critical water molecule in the binding site](1HSG (5).png)

> Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.

![An image identifying the critical water, both chains, the ligand, and D25 residues of both chains](1HSG (6).png)

## The Bio3D package

The bio3d package allows us to do all sorts of structural bioinformatics work in R

Lets start with how it can read these PDB files:



```{r}
library(bio3d)

pdb<- read.pdb("1hsg")
pdb
```
```{r}
attributes(pdb)
```
```{r}
head(pdb$atom)
```
```{r}
pdbseq(pdb)[25]
```
> Q7: How many amino acid residues are there in this pdb object?

```{r}
sum(pdb$calpha)
```
This pdb object has 198 amino acid residues

> Q8: Name one of the two non-protein residues? 

HOH and MK1 are the two non-protein residues

> Q9: How many protein chains are in this structure?

```{r}
unique(pdb$atom$chain)
```
2 protein chains are in this structure

## Predicting functional motions of a single structure

```{r}
adk <- read.pdb("6s36")
adk
```
```{r}
# Perform flexiblity prediction
m <- nma(adk)
plot(m)
```
Write out a multi-model PDB file (trajectory) that we can use to make an animation of the predicted motions. 

```{r}
mktrj(m, file="adk_m7.pdb")
```

I can open this in Mol* to play the trajectory

## Comparative Structure Analysis of Adenylate Kinase

> Q10. Which of the packages above is found only on BioConductor and not CRAN?

Package "msa" only found on BioConductor

> Q11. Which of the above packages is not found on BioConductor or CRAN?: 

Package "Bio3d" not found on either BioConductor or CRAN

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

True


```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

> Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

There are 214 amino acids that make up the sequence according to the output seq above.

```{r}
b <- blast.pdb(aa)
```
```{r}
hits <- plot(b)
```
```{r}
head(hits$pdb.id)
```
```{r}
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```
```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
pdbs
```
```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment (Will not format to pdf - only code shown)
##plot(pdbs, labels=ids)

```
```{r}
#Annotate PDB Structures
anno <- pdb.annotate(ids)
unique(anno$source)
anno
```

**Principle Component Analysis**

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```
```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```
```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```
```{r}
# Normal Mode Analysis of all structures
modes <- nma(pdbs)
plot(modes, pdbs, col=grps.rd)
```
> Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

I notice that this plot may show distinct configurations of the protein of interest based on the activities of the black and colored lines. These lines are different as the colored lines represent more peaks and troughs compared to the black line which exhibits much less fluctuations and peak variability. This difference could be in leu of the greater conformations present in one region versus the other as we saw previously where certain structures of the protein are flexible and can change its conformation and possibly alter its activity as well. 

